chainer/functions/loss/discriminative_loss.py

```diff
@@ -77,65 +76,85 @@ def __call__(self, embeddings, labels):
               ``gamma``
 
         """
-        assert (self.max_embedding_dim == embeddings.shape[1])
 
-        l_dist = 0.0
-        count = 0
+        shape = embeddings.shape
+        assert len(shape) == 4
+
         xp = backend.get_array_module(embeddings)
 
-        emb = embeddings[None, :]
-        emb = broadcast_to(emb, (emb.shape[1],
-                                 emb.shape[1],
-                                 emb.shape[2],
-                                 emb.shape[3],
-                                 emb.shape[4]))
-        ms = []
-        for c in range(self.max_embedding_dim):
-            # Create mask for instance
-            mask = xp.expand_dims(labels == c + 1, 1)
-            ms.append(mask)
-        if hasattr(xp, 'stack'):
-            ms = xp.stack(ms, 0)
-        else:
-            # Old numpy does not have numpy.stack.
-            ms = xp.concatenate([xp.expand_dims(x, 0) for x in ms], 0)
-        mns = c_sum(emb * ms, axis=(3, 4))
-        mns = mns / xp.maximum(xp.sum(ms, (2, 3, 4))[:, :, None], 1)
-        mns_exp = mns[:, :, :, None, None]
+        unique_label_idx = xp.unique(labels)
+        # Remove background label
+        unique_label_idx = unique_label_idx[unique_label_idx >= 0]
+
+        var_loss = xp.zeros((shape[0],))
```

Comment:
How about improving readability by
```
b = embeddings.shape[0]
```
and then using `b` throughout the code to denote the batch size?
```
xp.zeros(b)
```

Reply:
I think `var_loss` should have the same dtype as `embeddings`.